dask/dataframe/io/parquet/core.py

```diff
@@ -0,0 +1,472 @@
+    engine : {'auto', 'fastparquet', 'pyarrow'}, default 'auto'
+        Parquet library to use. If only one library is installed, it will use
+        that one; if both, it will use 'fastparquet'.
+    compression : string or dict, optional
+        Either a string like ``"snappy"`` or a dictionary mapping column names
+        to compressors like ``{"name": "gzip", "values": "snappy"}``. The
+        default is ``"default"``, which uses the default compression for
+        whichever engine is selected.
+    write_index : boolean, optional
+        Whether or not to write the index. Defaults to True *if* divisions are
+        known.
+    append : bool, optional
+        If False (default), construct data-set from scratch. If True, add new
+        row-group(s) to an existing data-set. In the latter case, the data-set
+        must exist, and the schema must match the input data.
+    partition_on : list, optional
+        Construct directory-based partitioning by splitting on these fields'
+        values. Each dask partition will result in one or more datafiles,
+        there will be no global groupby.
+    storage_options : dict, optional
+        Key/value pairs to be passed on to the file-system backend, if any.
+    write_metadata_file : bool, optional
+        Whether to create the special "_metadata" file,
+    compute : bool, optional
+        If True (default) then the result is computed immediately. If False
+        then a ``dask.delayed`` object is returned for future computation.
+    **kwargs
+        Extra options to be passed on to the specific backend.
+
+    Examples
+    --------
+    >>> df = dd.read_csv(...)  # doctest: +SKIP
+    >>> to_parquet('/path/to/output/', df, compression='snappy')  # doctest: +SKIP
+
+    See Also
+    --------
+    read_parquet: Read parquet data to dask.dataframe
+    """
+    partition_on = partition_on or []
+    if isinstance(partition_on, string_types):
+        partition_on = [partition_on]
+
+    if set(partition_on) - set(df.columns):
+        raise ValueError(
+            "Partitioning on non-existent column. "
+            "partition_on=%s ."
+            "columns=%s" % (str(partition_on), str(list(df.columns)))
+        )
+
+    if compression != "default":
+        kwargs["compression"] = compression
+    elif "snappy" in compress:
+        kwargs["compression"] = "snappy"
+
+    if isinstance(engine, str):
+        engine = get_engine(engine)
+
+    fs, fs_token, _ = get_fs_token_paths(
+        path, mode="wb", storage_options=storage_options
+    )
+    # Trim any protocol information from the path before forwarding
+    # ideally, this should be done as a method of the file-system
+    path = infer_storage_options(path)["path"]
+
+    if write_index:
+        df = df.reset_index()
+
+    # create parquet metadata, includes loading of existing stuff is appending
+    meta, filenames, kwargs2 = engine.create_metadata(df, fs, path, append)
+    kwargs.update(kwargs2)
```

Comment:
I'm not sure I understand the difference between create_metadata and write_metadata.  I wouldn't have expected us to write anything until we had the partial metadata bits from the write_partition calls.

Reply:
This is the change I was trying to get across here. Metadata processing like dealing with append and deciding on the set of filenames to assign to the parts must happen before creating the writing tasks for each partition; but actually writing out a metadata file will only happen once the writes are done (if at all).